### PR TITLE
bump-formula-pr: enable same-repo (no-fork) Pull Requests

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

To build bottles with GitHub actions, we need to create PRs from branches that live in the same tap. Here I add `--no-fork` option that instructs `bump-formula-pr` to create branch in `origin` rather than in a forked repo.

Tested at https://github.com/Linuxbrew/homebrew-xorg/pull/581

### CODE REVIEW

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 1]
```diff
@@ -298,27 +300,32 @@ def bump_formula_pr
...
-          # If the repository is private, forking might be disabled.
-          # Create branches in the repository itself instead.
+        if args.no_fork?
           remote_url = Utils.popen_read("git remote get-url --push origin").chomp
           username = formula.tap.user
-        rescue *GitHub.api_errors => e
-          formula.path.atomic_write(backup_file) unless args.dry_run?
-          odie "Unable to fork: #{e.message}!"
+        else
+          begin
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Maybe this could all be extracted to another function to make use of e.g. early returns?

**Comment by issyl0** [comment]
This needs a rebase now #6718 has been merged.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 2]
```diff
@@ -394,6 +380,23 @@ def bump_formula_pr
     end
   end
 
+  def forked_repo_info(tap_full_name)
+    response = GitHub.create_fork(tap_full_name)
+  rescue GitHub::AuthenticationFailedError, *GitHub.api_errors => e
+    formula.path.atomic_write(backup_file) unless args.dry_run?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: `args.dry_run` will never be `true` here.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 3]
```diff
@@ -394,6 +380,23 @@ def bump_formula_pr
...
+    odie "Unable to fork: #{e.message}!"
+  else
+    # GitHub API responds immediately but fork takes a few seconds to be ready.
+    sleep 1 until GitHub.check_fork_exists(tap_full_name)
+    if system("git", "config", "--local", "--get-regexp", "remote\..*\.url", "git@github.com:.*")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]:
```suggestion
    remote_url =  if system("git", "config", "--local", "--get-regexp", "remote\..*\.url", "git@github.com:.*")
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 4]
```diff
@@ -394,6 +380,23 @@ def bump_formula_pr
...
+      remote_url = response.fetch("ssh_url")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]:
```suggestion
      response.fetch("ssh_url")
```

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 5]
```diff
@@ -394,6 +380,23 @@ def bump_formula_pr
...
+    else
+      remote_url = response.fetch("clone_url")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]:
```suggestion
      response.fetch("clone_url")
```

File: Library/Homebrew/utils/github.rb [hunk 6]
```diff
@@ -340,6 +340,15 @@ def create_fork(repo)
     open_api(url, data: data, scopes: scopes)
   end
 
+  def check_fork_exists(repo)
+    username = api_credentials[1]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]:
```suggestion
    _, username = api_credentials
```

File: Library/Homebrew/utils/github.rb [hunk 7]
```diff
@@ -340,6 +340,15 @@ def create_fork(repo)
...
+    reponame = repo.split("/")[1]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]:
```suggestion
    _, reponame = repo.split("/")
```

File: Library/Homebrew/utils/github.rb (hunk 6 above)
Comment (maxim-belkin) [reply Library/Homebrew/utils/github.rb]: is there an advantage to using this syntax?

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 2 above)
Comment (maxim-belkin) [reply Library/Homebrew/dev-cmd/bump-formula-pr.rb]: ah, nice catch!

File: Library/Homebrew/utils/github.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/utils/github.rb]: Yes, it will avoid failing if it's not an array.

**Comment by MikeMcQuaid** [comment]
> #### Conflicting files
> `Library/Homebrew/dev-cmd/bump-formula-pr.rb`

Just to note this. Should be good to go after that (pending one final review).

**Comment by maxim-belkin** [comment]
Rebased the branch and it seems to be working as expected.
Tested here: https://github.com/Linuxbrew/homebrew-xorg/pull/588

**Comment by MikeMcQuaid** [comment]
Thanks again @fxcoudert!

**Comment by dawidd6** [comment]
I think you meant @maxim-belkin :)

**Comment by MikeMcQuaid** [comment]
Oops 😉! Edited!

**Comment by dawidd6** [comment]
This PR broke current standard workflow with forks. I'll shamelessly paste here the output that @chenrui333 got:
```
$ brew bump-formula-pr babel --url=https://registry.npmjs.org/@babel/core/-/core-7.7.4.tgz --sha256=3bb9a1d79c68a0c1c41885b86471d306b99efefabb2def80ea3a511d3476f048 --verbose
==> replace /https:\/\/registry\.npmjs\.org\/@babel\/core\/\-\/core\-7\.7\.2\.tgz/ with "https://registry.npmjs.org/@babel/core/-/core-7.7.4.tgz"
==> replace "e2e9745819df1a040265933785536c2b3317661bc635c63e98497e4882ae6a39" with "3bb9a1d79c68a0c1c41885b86471d306b99efefabb2def80ea3a511d3476f048"
/usr/local/bin/brew audit /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/babel.rb
Error: Not Found
```

**Comment by maxim-belkin** [comment]
I see

brew audit /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/babel.rb
Error: Not Found


How is it related to GitHub.check_fork_exists(tap_full_name)?

**Comment by maxim-belkin** [comment]
The only change that should be noticeable after this PR is that now `bump-formula-pr` will not attempt to push to `origin` without `--no-fork` and _vice a versa_: with `--no-fork` it will not attempt to create a fork.

@dawidd6 and @chenrui333, could you please post here a repro next time you encounter an error? I see that `babel` has been updated...

**Comment by dawidd6** [comment]
The repro is actually pretty simple, try to `brew bump-formula-pr` anything without `--no-fork` option passed and the error will appear. Maybe it has something to do with:
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/github.rb#L82-L99
This function is used in `check_fork_exists(repo)` function that you introduced in this PR, but I only have `HOMEBREW_GITHUB_API_TOKEN` set and it seems like that function depends on `HOMEBREW_GITHUB_API_USERNAME` being set, yet I don't have the need to set it.

**Comment by maxim-belkin** [comment]
Ah, I think I know what the problem is. I’ll look into it later today.

**Pull request merged**